File: src/do_join.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Low resource full outer join

Hacky approach, but necessary because of file sizes
Please don't judge :)

Plan of attack
===========================
- Get new data from stdin
- Read existing table as working/out.csv
- Write concatenated lines to working/tmp.csv
- Copy tmp.csv to out.csv

Created on Mon Oct 28 19:08:07 2019
@author: Paul W. Egeler, M.S.
"""
import sys
import re
import shutil
import os.path

# Get input file
work_dir, file_name = os.path.split(sys.argv[1])

# Open files
temp   = open(os.path.join(work_dir, 'tmp.csv'), 'w+')
work   = open(os.path.join(work_dir, 'out.csv'), 'r')
infile = open(os.path.join(work_dir, file_name), 'r')

# Header
header = work.readline().strip() + ','
header += re.search(r'^sorted_(.*?)_', file_name).group(1) + '\n'
temp.write(header)

# Preload first line from stdin
candidate = infile.readline().strip().split()

# Loop over working file
for line in work:
    line = line.strip()
    temp.write(line + ',')
    if line.split(',', 4)[:4] == candidate[:4]:
        temp.write(candidate[-1])
        candidate = infile.readline().strip().split()
    temp.write('\n')

# Close file connections
temp.close()
work.close()
infile.close()

# At last, copy tmp into out
shutil.move(os.path.join(work_dir, 'tmp.csv'),
            os.path.join(work_dir, 'out.csv'))

File: src/do_join.cpp
#include <fstream>
#include <string>
#include <regex>
#include <filesystem>
#include <vector>
#include <boost/algorithm/string.hpp>

using namespace std;
namespace fs = filesystem;

bool compare_keys(const vector<string>& v1, const vector<string>& v2)
{
  for ( int i=0; i < 4; i++ )
    if ( v1[i] != v2[i] )
      return false;
  return true;
}

vector<string> tokenize_line(const string& line, const string& delim)
{
  vector<string> fields;
  boost::split(fields, line, boost::is_any_of(delim));
  return fields;
}

int main(int argc, char *argv[])
{
  fs::path p = argv[1];
  string stem = p.stem().string();

  // Open the files
  ifstream infile(p);
  ifstream work(p.parent_path() / "out.csv");
  ofstream temp(p.parent_path() / "tmp.csv");

  // Accession number
  regex re("^sorted_(.*?)_.*");
  smatch sm;
  regex_match(stem, sm, re);

  // Dim strings
  string line, candidate;

  // Header
  getline(work, line);
  temp << line << "," << sm[1] << "\n";

  // Preload first line from infile
  getline(infile, candidate);
  vector<string> candidate_tokenized = tokenize_line(candidate, "\t");

  // Loop over working file
  while ( getline(work, line) )
  {
    temp << line << ",";
    vector<string> line_tokenized = tokenize_line(line, ",");

    if ( compare_keys(line_tokenized, candidate_tokenized) )
    {
      temp << candidate_tokenized[6];
      getline(infile, candidate);
      candidate_tokenized = tokenize_line(candidate, "\t");
    }

    temp << "\n";
  }

  // Close file connections
  temp.close();
  work.close();
  infile.close();

  // Overwrite old working file
  fs::rename(p.parent_path() / "tmp.csv",
             p.parent_path() / "out.csv");

  return 0;
}

